compiler/ast/ast.go

```diff
@@ -71,58 +97,103 @@ type Conditional struct {
 // a function call has the standard semantics where it takes one or more arguments
 // and returns a result.
 type Call struct {
-	Kind  string `json:"kind" unpack:""`
-	Name  string `json:"name"`
-	Args  []Expr `json:"args"`
-	Where Expr   `json:"where"`
+	Kind    string `json:"kind" unpack:""`
+	Name    string `json:"name"`
+	NamePos int    `json:"name_pos"`
+	Args    []Expr `json:"args"`
+	Rparen  int    `json:"rparen"`
+	Where   Expr   `json:"where"`
+}
+
+func (c *Call) Pos() int { return c.NamePos }
+
+func (c *Call) End() int {
+	if c.Where != nil {
+		return c.Where.End()
+	}
+	return c.Rparen
 }
 
 type Cast struct {
-	Kind string `json:"kind" unpack:""`
-	Expr Expr   `json:"expr"`
-	Type Expr   `json:"type"`
+	Kind   string `json:"kind" unpack:""`
+	Expr   Expr   `json:"expr"`
+	Type   Expr   `json:"type"`
+	Rparen int    `json:"rparen"`
 }
 
+func (c *Cast) Pos() int { return c.Type.Pos() }
+func (c *Cast) End() int { return c.Rparen + 1 }
+
 type IndexExpr struct {
-	Kind  string `json:"kind" unpack:""`
-	Expr  Expr   `json:"expr"`
-	Index Expr   `json:"index"`
+	Kind   string `json:"kind" unpack:""`
+	Expr   Expr   `json:"expr"`
+	Index  Expr   `json:"index"`
+	Rbrack int    `json:"rbrack"`
 }
 
+func (i *IndexExpr) Pos() int { return i.Expr.Pos() }
+func (i *IndexExpr) End() int { return i.Rbrack + 1 }
+
 type SliceExpr struct {
-	Kind string `json:"kind" unpack:""`
-	Expr Expr   `json:"expr"`
-	From Expr   `json:"from"`
-	To   Expr   `json:"to"`
+	Kind   string `json:"kind" unpack:""`
+	Expr   Expr   `json:"expr"`
+	From   Expr   `json:"from"`
+	To     Expr   `json:"to"`
+	Rbrack int    `json:"rbrack"`
 }
 
+func (s *SliceExpr) Pos() int { return s.Expr.Pos() }
+func (s *SliceExpr) End() int { return s.Rbrack + 1 }
+
 type Grep struct {
 	Kind    string `json:"kind" unpack:""`
 	Pattern Expr   `json:"pattern"`
 	Expr    Expr   `json:"expr"`
+	Left    int    `json:"left"`
```

Comment:
If I understand correctly, this is the position of the "grep" keyword, so maybe call this KeywordPos and put it between Kind and Pattern?